contrib/thrifty/src/python/pants/contrib/thrifty/java_thrifty_gen.py

```diff
@@ -0,0 +1,100 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import logging
+import os
+
+from pants.backend.jvm.targets.java_library import JavaLibrary
+from pants.backend.jvm.tasks.nailgun_task import NailgunTaskBase
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.java.jar.jar_dependency import JarDependency
+from pants.task.simple_codegen_task import SimpleCodegenTask
+from twitter.common.collections import OrderedSet
+
+from pants.contrib.thrifty.java_thrifty_library import JavaThriftyLibrary
+
+
+logger = logging.getLogger(__name__)
+
+
+class JavaThriftyGen(NailgunTaskBase, SimpleCodegenTask):
+
+  gentarget_type = JavaThriftyLibrary
+
+  @classmethod
+  def register_options(cls, register):
+    super(JavaThriftyGen, cls).register_options(register)
+
+    def thrifty_jar(name):
+      return JarDependency(org='com.microsoft.thrifty', name=name, rev='0.4.3')
+
+    cls.register_jvm_tool(register,
+                          'thrifty-runtime',
+                          classpath=[thrifty_jar(name='thrifty-runtime')])
+    cls.register_jvm_tool(register,
+                          'thrifty-compiler',
+                          classpath=[thrifty_jar(name='thrifty-compiler')])
+
+  def __init__(self, *args, **kwargs):
+    super(JavaThriftyGen, self).__init__(*args, **kwargs)
+
+  def synthetic_target_type(self, target):
+    return JavaLibrary
+
+  def synthetic_target_extra_dependencies(self, target, target_workdir):
+    deps = OrderedSet(self.resolve_deps([self.get_options().thrifty_runtime]))
+    deps.update(target.dependencies)
+    return deps
+
+  def synthetic_target_extra_exports(self, target, target_workdir):
+      return self.synthetic_target_extra_dependencies(target, target_workdir)
```

Comment:
indent is odd.

But also, generally it's not a good idea to export _everything_ that you depend on. Not a huge deal, but would recommend starting conservatively and exporting only the "required/default" dependency (the `thrifty-runtime`). Not a blocker, but relatively hard to fix later, so good to figure out early.

Reply:
This is probably a gap in my understanding about what should be exported. In the case where `java_thrifty_library` targets depend on other `java_thrifty_library` targets, do we need to export that IDL target chain, or just the `thrifty-runtime` dep?

Reply:
It depends what the generated code looks like, but the answer is "sometimes".

If you look at `java_thrift_library`, it takes an `exports` argument that allows users to say "I depend on 10 things, but only export 4 of them", similar to what you would do for non-generated code.

This is only really relevant when you have `strict_deps` enabled for your targets, but if you don't care either way, I'd recommend going conservative and not exporting much: a dependent can always just add an explicit dependency, and it's much easier to loosen the constraints later than it is to tighten them.